022-feature-gates.md

```diff
@@ -0,0 +1,105 @@
+
+An example of this use case might be the introduction of the separate control plane listener.
+The listener it self can be added by default.
+But the feature gate would control whether it is actually used or not.
+At first, it will be disabled by default allowing smooth upgrades without going through every single Strimzi version.
+But users who want to use this feature will be able to enable it and benefit from it.
+Later it would be enabled by default - most users will use it, but those upgrading from much older versions without the control plane listener would still be able to disable it, upgrade and enable it.
+After sufficient amount of releases, the feature gate will be removed and the separate control plane listener will be used by all the time.
+
+#### Introducing experimental or unstable APIs
+
+The feature gate will be disabled by default to allow testing of new / experimental functionality.
+Features might be seen as experimental because of being considered risky (with high probability of bugs and issues) or because the feature / behavior might change in the future.
+As the new feature matures and becomes less experimental, it might be enabled by default.
+Finally, when the feature is seen as stable and ready for _production_, the feature gate might be removed and the feature enabled by default without any option to disable it.
+
+And example of this use-cases might be for example using a feature gate for the StatefulSet removal or possibly some features related to KIP-500 / ZooKeeper removal.
+
+#### Enabling / disabling a feature
+
+In some cases, different users might have different preferences and it might be hard to find a middle ground which would make everyone happy and satisfied.
+In such case, the feature gates might be used to modified the behavior of the operator and enable / disable certain features.
+In use case like this, the default state of the feature gate should be based on what we expect to be more common or set to keep the original behavior to maintain backwards compatibility.
+Feature gates used for this use cases might be never removed.
+
+An example of this use case might be feature gates to disable automatic creation of Pod Disruption Budgets or Network Policies.
+
+### Implementation
+
+Initially the feature gates support should be implemented in Cluster Operator.
+But the same logic can be also applied to User and Topic Operators.
+The feature gates will be configured in an environment variable `STRIMZI_FEATURE_GATES`.
+Since the feature gates might be enabled or disabled by default, the configuration needs to consist of coma-separated key-value pairs.
+The key will be the name of the feature gate and the value will be a boolean defining the state -  either enabling it (`true`) or disabling it (`false`).
+For example `PodDisruptionBudgets=false,NetworkPolicies=false,ControlPlaneListener=true`.
```

Comment:
Alternative: `STRIMZI_FEATURE_GATES=-PodDisruptionBudgets,-NetworkPolicies,+ControlPlaneListener`. I don't have a strong opinion though.

I assume it's an error for a name to appear more than once?